### ISSUE TITLE
Allow predicate pushdown for Avro IO

### ISSUE THREAD

**Issue by anish749** [spotify/scio#1974]
`AvroSpecificRecordIO` needs to deserialise all records and all fields in all records before it can apply any filtering. Though the no of columns being deserialised can be reduced by specifying a lighter schema, the same can't be done rows. We can push down a function and apply the function before deserialisation, to make reading input more performant.

Directly using `AvroGenericRecordIO` is inefficient because the default `Coder[GenericRecord]` is inefficient.

More detailed examples, numbers with internal datasets [here](https://docs.google.com/document/d/1m5M0JgjrmRca8qWtzsJa5lJoCGBqRqRcv5h0JaE3NYE/edit#heading=h.q4tiuatwvivr)

I tried out a rough (incomplete) version in #1973

**Comment by nevillelyh** [comment]
How do we generate the lighter schema though? Does the user have to handcraft one (inconvenient), do we allow some stringly-typed field selector similar to BigQuery storage API and generate programmatically, or do we analyze the `parseFn` via macro and generate it, similar to my Parquet [macros](https://github.com/nevillelyh/parquet-avro-extra)?

**Comment by nevillelyh** [comment]
Made this quick test of using a slim reader schema, and it's 2x as fast.
https://github.com/nevillelyh/parquet-avro-extra/blob/neville/avro/parquet-avro-examples/src/main/scala/AvroTest.scala

I had to tweak parquet-avro-extra `Projection` to not randomize record names but other than that it's very straightforward. Will try it with a proper Scio job to benchmark further.

**Comment by anish749** [comment]
My initial thought was to have a handcrafted light schema OR use stringly typed funciton like BQ Storage.

I now have a vague thought that with Scio 0.8 and the `Schema` type class, it might be easier to have some checks at compile time.
We can allow the user to write down a function, and at compile time check if the `Schema` of the inputs of the predicate is a subset of the `Schema[T]` where `T <: SpecificRecord`. In this way we can have some safety while still being able to use the predicate before deserialization.

Eg:
```scala
val coll: SCollection[WideRecordEndSong] = sc.avroFile(path, predicate = (country:String) => country == "US") )
```
and the function signature being:
```scala
def avroFile[L: Schema, T <: SpecificRecord: Schema](path:String, predicate: L => Boolean)(implicit evidence: SchemaSubset[L, T]) 

@implicitNotFound("Cannot prove {L} is derivable from {T}")
trait SchemaSubset[L, T]{
   implicit def subset(implicit ls: Schema[L], ts: Schema[T]): SchemaSubset[L, T] = macro ....
   // The above macro will fail to derive an instance of SchemaSubset if ls is not derivable from ts
}
```

Having a `SchemaSubset` typeclass would also help solving #1954 easier.

**Comment by nevillelyh** [comment]
Looks like there are 2 orthogonal problems we're trying to solve:

1. predicate push down and avoid deserializing the entire row
1. column projection to copy only subset of the columns

1 most likely will require a non-trivial macro which I'm a little hesitant given the complexity & ROI.
2 is probably doable if we require a user specified `case class T`and and either codegen `GenericRecord => T`, reuse something like [shapeless-datatype](https://github.com/nevillelyh/shapeless-datatype), or the `Row => T` conversion from Scio Beam SQL integration.

Maybe hack some lightweight solution for 2 and measure first? Feel like it's more common in ETL pipelines than 1.

**Comment by anish749** [comment]
I don't think I understood this clearly.
For no 2, if we just remove unnecessary columns from the reader schema of type `T <: SpecificRecord`, it would work, right?. So if I am interested in only 2 columns for my pipeline, I create a `schema.avsc` with only two fields, and read as `SpecificRecord`.  Don't we already have that?

**Comment by nevillelyh** [comment]
It'll work to some extent:
- We need to patch upstream `AvroIO`, `AvroSource` to allow reader schema in addition to `Class[T <:< SpecificRecord]`, and I ran into some complication like cast errors during conversion.
- The projected `SpecificRecord` is incomplete, and if any unprojected column is nullable, will fail `SerializableUtils.ensureSerializable`. So we'll have to plug in a SR to case class function anyway.

**Comment by regadas** [comment]
> column projection to copy only subset of the columns

`case class` conversion based on `Schema` type class is already possible but keep in mind that it's still experimental. Also for avro we have `SpecificRecord => T`

> predicate push down and avoid deserializing the entire row

this is interesting! For `GenericRecord` #1973 seems to do the job but for `SpecificRecord` it would indeed be nice to supply different read schema.

**Comment by regadas** [comment]
@nevillelyh curious on why would we need to patch upstream? what am I missing?

**Comment by nevillelyh** [comment]
@regadas upstream `AvroIO` doesn't allow a different reader schema when reading `SpecificRecord` and will infer schema from the compiled class. It's actually more complicated. Even if I can read `SpecificRecord` with a light reader schema, the un-projected non-nullable fields will cause NPE in during ser/de in mutation detector, and we'll have to force user to project those fields out with an additional map `fn`, similar to the Parquet Avro projection. I'm not sure the performance improvement is worth the effort.

**Comment by regadas** [comment]
@nevillelyh So your point is that for a given compiled `LargeSpecificRecord` and with the eventual possibility of supplying a lighter read schema we would need to supply a map `fn` afterwards. Seems about right! But my point is why is this more advantageous than just supplying a compiled record with a lighter schema?

the above example works.

**Comment by nevillelyh** [comment]
"compiled record with a lighter schema" => user has to do the extra work to create and compile such schema, a pretty big overhead that many would avoid.

**Comment by regadas** [comment]
couldn't users simply also use `@AvroType.fromSchema`?

**Comment by nevillelyh** [comment]
Yes, but they still have to dump, copy/paste schema and manually trim fields. Plus `AvroType` adds the Avro => case class conversion overhead that probably offsets the speed up gained in projection.

**Comment by regadas** [comment]
Indeed, case class can be a potential issue!

Another potential issue with the previous approach is that `fn: SpecificRecord => T` needs to be in sync with the supplied `LightSchema`.  `fn` might be accessing fields that will return `null` if schema changes.

**Comment by regadas** [comment]
Closed via #1973. `GenericRecordParseIO` seems to offer a good solution to this for the time being.
